ESP8266: implements possibility to decide between non-blocking/blocking connect.

VeijoPesonen opened the pull request:
### Description
Checks for already existing connection when trying to set credentials for a new connection or when trying to call connect. This is done to avoid the situation where an already established connection would get disrupted.

Implements possibility to decide between non-blocking/blocking connect.

### Pull request type
    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers
@SeppoTakalo 
@michalpasztamobica
@marcuschangarm 
@karsev 
@teetak01

Commit:
ESP8266: makes connect() connection status aware

michalpasztamobica commented on components/wifi/esp8266-driver/ESP8266Interface.cpp:
```diff
@@ -716,4 +724,27 @@ void ESP8266Interface::proc_oob_evnt()
+        case NSAPI_STATUS_GLOBAL_UP:
+            ret = NSAPI_ERROR_IS_CONNECTED;
+            break;
+        default:
```
I noticed that it is possible for ESP8266 to also set NSAPI_STATUS_ERROR_UNSUPPORTED (in _oob_watchdog_reset() ). Should we add a case for it, in order not hide it behind the DEVICE_ERROR?

VeijoPesonen replied:
Maybe, but what is a watchdog reset if not a device error. There is an OOB handler for the watchdog reset and a network status update is send to the application. I think device error should be fine.

michalpasztamobica replied:
Right, I actually wonder why getting a watchdog reset causes the connection status to be UNSUPPORTED... Seems like there is just no better status for it and it will likely be set to something meaningful after the reset takes place. 
I agree - it is better leave the code as is :)

VeijoPesonen replied:
Got an idea that it would be better to call MBED_ERROR if the modem crashes and watchdog reset occurs. Better to reveal the crash as soon as possible.

michalpasztamobica replied:
Fully agree. It is not relevant for this PR, so I added it to internal JIRA ticket ONME-4141, so we don't forget.

VeijoPesonen replied:
Created an issue #9425

michalpasztamobica reviewed (commented):
Just a single question to the switch-case statement. Otherwise a very good and helpful change :)

ciarmcom commented:
@VeijoPesonen, thank you for your changes.
@SeppoTakalo @michalpasztamobica @marcuschangarm @karsev @ARMmbed/mbed-os-ipcore @ARMmbed/mbed-os-maintainers please review.

0xc0170 commented:
Ci started

Commit:
ESP8266: turns connect fully to non-blocking mode

Blocking mode connect is not supported anymore when this patch is
applied

0xc0170 commented:
> @SeppoTakalo
@marcuschangarm
@karsev

Any of you to review this patch ? It's ready for integration

Commit:
ESP8266: provides blocking/non-blocking connect

Implements NetworkInterface::set_blocking() and implements the
functionality to distinguish between the two in connect()

VeijoPesonen commented:
Adding the non-blocking connect itself. @kjbracey-arm, @michalpasztamobica, @SeppoTakalo  please re-review.

0xc0170 commented:
CI started

kjbracey commented on components/wifi/esp8266-driver/ESP8266Interface.cpp:
```diff
@@ -194,11 +232,33 @@ int ESP8266Interface::connect()
+    }
+
+    if (_if_blocking) {
+        _if_connected.wait();
```
This needs to be a proper condition test loop. Reread the CV documentation

    while (connection_status != CONNECTED) {
        _if_connected.wait();
    }

VeijoPesonen replied:
Fixed

kjbracey commented on components/wifi/esp8266-driver/ESP8266Interface.cpp:
```diff
@@ -167,9 +179,35 @@ void ESP8266Interface::_oob2global_event_queue()
+    }
+
+    if (_esp.connect(ap_ssid, ap_pass) != NSAPI_ERROR_OK) {
+        _connect_event_id = _global_event_queue->call(callback(this, &ESP8266Interface::_connect_async));
```
As you will potentially be delaying the global event queue quite a long time here (15 second timeout), at least mitigate it by making this call again after the same timeout or longer. Don't take 100% of the thread time.

VeijoPesonen replied:
Fixed

VeijoPesonen commented on components/wifi/esp8266-driver/ESP8266Interface.cpp:
```diff
@@ -167,9 +179,36 @@ void ESP8266Interface::_oob2global_event_queue()
+
+    if (_esp.connect(ap_ssid, ap_pass) != NSAPI_ERROR_OK) {
+        // Postpone to give other stuff time to run
+        _connect_event_id = _global_event_queue->call_in(ESP8266_CONNECT_TIMEOUT, callback(this, &ESP8266Interface::_connect_async));
```
Delay added

VeijoPesonen commented on components/wifi/esp8266-driver/ESP8266Interface.cpp:
```diff
@@ -194,11 +233,32 @@ int ESP8266Interface::connect()
+    _cmutex.lock();
+
+    MBED_ASSERT(!_connect_event_id);
+    _connect_event_id = _global_event_queue->call(callback(this, &ESP8266Interface::_connect_async));
```
No need for delay here as this is the first attempt.

0xc0170 commented:
CI aborted until review is completed

0xc0170 commented:
CI started (travis restarted as well)

cmonr commented:
CI job restarted: `jenkins-ci/mbed-os-ci_greentea-test`

Failure appeared unrelated to PR.

0xc0170 commented:
CI restarted
